api/unstable/crl.h

```diff
@@ -169,3 +169,69 @@ S2N_API int s2n_crl_lookup_set(struct s2n_crl_lookup *lookup, struct s2n_crl *cr
  * @return S2N_SUCCESS on success, S2N_FAILURE on failure.
  */
 S2N_API int s2n_crl_lookup_ignore(struct s2n_crl_lookup *lookup);
+
+struct s2n_cert_validation_info;
+
+/**
+ * A callback which can be implemented to perform additional validation on received certificates.
+ *
+ * The cert validation callback is invoked after receiving and validating the peer's certificate chain. The callback
+ * can be used by clients to validate server certificates, or by servers to validate client certificates in the case of
+ * mutual auth. Note that any validation performed by applications in the callback is in addition to the certificate
+ * validation already performed by s2n-tls.
+ *
+ * Applications can use either of the following APIs from within the callback to retrieve the peer's certificate chain
+ * and perform validation before proceeding with the handshake:
+ *  - `s2n_connection_get_peer_cert_chain()`
+ *  - `s2n_connection_get_client_cert_chain()`
+ *
+ * If the validation performed in the callback is successful, `s2n_cert_validation_accept()` MUST be called to allow
+ * `s2n_negotiate()` to continue the handshake. If the validation is unsuccessful, `s2n_cert_validation_reject()`
+ * MUST be called, which will cause `s2n_negotiate()` to error. The behavior of `s2n_negotiate()` is undefined if
+ * neither `s2n_cert_validation_accept()` or `s2n_cert_validation_reject()` are called.
+ *
+ * Applications can specify an error code when calling `s2n_cert_validation_reject()` by setting an error code field
+ * on the context for the `s2n_connection`, via the `s2n_connection_set_ctx()` API. The error code can be set from the
+ * callback by retrieving the context via `s2n_connection_get_ctx()`, and then retrieved again later when processing
+ * the `s2n_negotiate()` failure.
```

Comment:
Here it kind of sounds like you're telling applications to do this, instead of giving advice on how they could do this if they needed it. Like maybe say that s2n-tls will report a protocol error with an error indicating the application as the source of the validation, but that if application need specific custom error codes they can use the context.